examples/stable_diffusion_xl/gm/models/trainer_factory.py

```diff
@@ -64,14 +70,65 @@ def __init__(
         self.enable_conditioner = enable_conditioner
         self.enable_first_stage_model = enable_first_stage_model
 
+        self.timestep_bias_strategy = timestep_bias_strategy
+        self.timestep_bias_multiplier = timestep_bias_multiplier
+        self.timestep_bias_begin = timestep_bias_begin
+        self.timestep_bias_end = timestep_bias_end
+        self.timestep_bias_portion = timestep_bias_portion
+        self.timestep_bias_weighting = Tensor(self.generate_timestep_weights(model.sigma_sampler.num_idx))
+
+    def generate_timestep_weights(self, num_timesteps):
+        weights = np.ones(num_timesteps)
+
+        # Determine the indices to bias
+        num_to_bias = int(self.timestep_bias_portion * num_timesteps)
+
+        if self.timestep_bias_strategy == "later":
+            bias_indices = slice(-num_to_bias, None)
+        elif self.timestep_bias_strategy == "earlier":
+            bias_indices = slice(0, num_to_bias)
+        elif self.timestep_bias_strategy == "range":
+            # Out of the possible 1000 timesteps, we might want to focus on eg. 200-500.
+            range_begin = self.timestep_bias_begin
+            range_end = self.timestep_bias_end
+            if range_begin < 0:
+                raise ValueError(
+                    "When using the range strategy for timestep bias, you must provide a beginning timestep greater or equal to zero."
+                )
+            if range_end > num_timesteps:
+                raise ValueError(
+                    "When using the range strategy for timestep bias, you must provide an ending timestep smaller than the number of timesteps."
+                )
+            bias_indices = slice(range_begin, range_end)
+        else:  # 'none' or any other string
+            return weights
+        if self.timestep_bias_multiplier <= 0:
+            return ValueError(
+                "The parameter --timestep_bias_multiplier is not intended to be used to disable the training of specific timesteps."
+                " If it was intended to disable timestep bias, use `--timestep_bias_strategy none` instead."
+                " A timestep bias multiplier less than or equal to 0 is not allowed."
+            )
+
+        # Apply the bias
+        weights[bias_indices] *= self.timestep_bias_multiplier
+
+        # Normalize
+        weights /= weights.sum()
+
+        return weights
+
     def construct(self, x, *tokens):
         # get latent target
         if self.enable_first_stage_model:
             x = self.first_stage_model.encode(x)
         x = self.scale_factor * x
 
         # get noise and sigma
-        sigmas = self.sigma_sampler(x.shape[0])
+        if self.timestep_bias_strategy == "none":
```

Comment:
Why `"none"` and not `None`?

Reply:
the type of arg "timestep_bias_strategy" set to str and choose from ["earlier", "later", "range", "none"]

Reply:
remove `default="none"` from the parser and the value will be `None` by default when no arguments are specified. Don't use `"none"` as a substitution of `None`.